tensorflow/compiler/xla/service/gpu/horizontal_fusion.cc

```diff
@@ -0,0 +1,514 @@
+        outputs[i]->shape().element_type()) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+bool IsConsumerTheOnlyNonRootUser(const HloInstruction& instr,
+                                  const HloInstruction& consumer) {
+  return absl::c_all_of(instr.users(), [&](const HloInstruction* user) {
+    if (user->opcode() == HloOpcode::kGetTupleElement) {
+      // Skip GTE.
+      return IsConsumerTheOnlyNonRootUser(*user, consumer);
+    } else if (user == &consumer) {
+      // `user` is `consumer`.
+      return true;
+    } else if (user == user->parent()->root_instruction()) {
+      // Consumed by ROOT is always fine, since it is impossible to create
+      // cycles through ROOT.
+      return true;
+    } else {
+      return false;
+    }
+  });
+}
+
+// Returns whether `instr` is a profitable candidate to be horizontally fused.
+// Since the primary benefit of horizontal fusion comes from reducing the
+// kernel launch overhead, we want to exclude the instructions with
+// insignificant kernel launch overhead. In other words, we exclude instructions
+// if their computation latencies are longer than launch latencies. We estimate
+// the computation latency of a given instruction by its shapes and the
+// instruction count in its fused computation. We roughly observe that if a
+// fusion instruction has shapes smaller than `kShapeThreshold` and has fewer
+// instructions than `kInstrCountThreshold`, it is launch-latency-bound and
+// profitable by horizontal fusion.
+bool IsProfitableFusionCandidate(const HloInstruction& instr) {
+  CHECK(instr.opcode() == HloOpcode::kFusion);
+  constexpr int64 kShapeThreshold = 128 * 2048;
+  constexpr int64 kInstrCountThreshold = 30;
+  auto root = instr.fused_expression_root();
+
+  // Too large shapes are not easily profitable.
+  if (root->opcode() == HloOpcode::kTuple) {
+    // Since all output shapes are the same, use the first shape as the
+    // representative.
+    auto shape = root->operand(0)->shape();
+    if (ShapeUtil::ElementsIn(shape) > kShapeThreshold) {
+      return false;
+    }
+  } else {
+    auto shape = root->shape();
+    if (ShapeUtil::ElementsIn(shape) > kShapeThreshold) {
+      return false;
+    }
+  }
+
+  // Having too many instructions is not easily profitable.
+  if (instr.fused_instruction_count() > kInstrCountThreshold) {
+    return false;
+  }
+
+  return true;
+}
+
+// Returns whether `fusion_instr` has only row-major layouts.
+// The horizontal fusion excludes computations with non-row-major layouts,
+// because fusing computations with different layouts can result in uncoalesced
+// memory accesses and cause great performance overhead.
+bool HasOnlyRowMajorLayout(const HloInstruction& fusion_instr) {
+  CHECK(fusion_instr.opcode() == HloOpcode::kFusion);
+  auto instrs = fusion_instr.fused_instructions_computation()->instructions();
+  for (auto instr : instrs) {
+    if (instr->shape().layout().format() != DENSE) {
```

Comment:
I assume SPARSE is not really used (for GPU at least). Let me know if my understanding is not accurate.

If the format is SPARSE, we should return false too? as the backend cannot really support it?